Add collision and velocity systems

jeremyadamsfisher opened the pull request:
A little rough, but it works!

I refactored the `{Left,Right}Player` components into a single `Player` component with a type attribute...because I couldn't figure out how to query on `LeftPlayer || RightPlayer` 🤷‍♂️

I think perhaps we should ditch the relative coordinate system, because its getting unwieldy to map between them and do collision detection using absolute coordinates

Commit:
add collision and velocity systems

jryio commented on src/systems/collision.rs:
```diff
@@ -0,0 +1,21 @@
+) {
+    if let Ok((ball_transform, mut ball_velocity, _)) = ball_query.single_mut() {
+        // TODO: check if the next frame, instead of the current frame, is a collision
+        for (player_transform, _) in player_query.iter() {
+            if ball_transform
+                .translation
+                .distance(player_transform.translation)
+                < 100.0
```
Clever solution! When I run the collision detection, the ball collides with the paddle too far away. 

I was wondering if you came across this function from bevy

`bevy::sprite::collide_aabb::collide`
https://docs.rs/bevy/0.4.0/bevy/sprite/collide_aabb/fn.collide.html

It takes the `position` and `size` of objects `a` and `b` and does a side bounding box collision detection.

jryio commented on src/main.rs:
```diff
@@ -25,23 +30,22 @@ fn main() {
 }
 
-struct Ball;
+pub struct Velocity(Vec2);
```
What are you thoughts on my adding `speed` as a separate component? The ball moves a bit fast right now!

`Velocity` is a unit vector and `[-1, 1]` and speed is a positive scalar `[0, infinity)`

jryio commented on src/main.rs:
```diff
@@ -120,29 +129,24 @@ fn startup_system(
-            right_position.y += 1.0;
-        } else if key.pressed(KeyCode::Down) {
-            right_position.y -= 1.0;
+fn keyboard_input_system(mut players: Query<(&mut Position, &Player)>, key: Res<Input<KeyCode>>) {
```
If you wanted to query for `LeftPlayer` or `RightPlayer` using bevy, it would be wrapped the component with `Option<T>`

https://bevy-cheatbook.github.io/programming/queries.html

```rust
fn check_zero_health(
    // access entities that have `Health` and `Transform` components
    // get read-only access to `Health` and mutable access to `Transform`
    // optional component: get access to `Player` if it exists
    mut query: Query<(&Health, &mut Transform, Option<&Player>)>,
) {
    // get all matching entities
    for (health, mut transform, player) in query.iter_mut() {
        eprintln!("Entity at {} has {} HP.", transform.translation, health.hp);

        // center if hp is zero
        if health.hp <= 0.0 {
            transform.translation = Vec3::ZERO;
        }

        if let Some(player) = player {
            // the current entity is the player!
            // do something special!
        }
    }
}

```

jryio reviewed (commented):
@jeremyadamsfisher I am going to push a small commit that uses bevy's collision detection, let me know what you think

jryio commented on src/systems/velocity.rs:
```diff
@@ -0,0 +1,9 @@
+
+pub fn velocity_system(mut query: Query<(&mut Position, &Velocity)>) {
+    for (mut position, velocity) in query.iter_mut() {
+        position.x += velocity.0.x;
+        position.y += velocity.0.y;
```
Also I just realized that this is moving the sprites at percentage based positions rather than pixel based positions.

Meaning the ball will traverse the 60% of the screen in 1 second. I don't think we want this.

jryio committed:
use bevy collide_aabb for collision detection

1. Use system sets to ensure that order is preserved
2. Make PlayerType public to distinguish between types
3. Add bevy `Size` component to entities for collision_system
